### DISCUSSION TITLE
[Feedback Requested] Dynamic panel configuration using "Configuration from Query" & "Rows to Fields" transformations

### DISCUSSION THREAD

**Discussion by pkolyvas** [grafana/grafana#38029] [Dashboards, Panels & Visualizations]
Grafana 8.1 shipped with the ability to configure panels using query results and the ability to convert rows to fields. This discussion will serve as a place to solicit feedback for these two transformations as per issue #4929.

Related: https://github.com/grafana/grafana/pull/36548

We believe these are the foundational changes needed to create dynamic panel configurations and would love to hear how these features (or their documentation) are or aren't meeting your needs.

Documentation:
- [Config from query results](https://grafana.com/docs/grafana/next/panels/transformations/config-from-query/)
- [Rows to fields](https://grafana.com/docs/grafana/next/panels/transformations/rows-to-fields/)

**Comment by offermannu** [comment]
IMO this is a very powerful and important feature. But I find it difficult to use because of the lack of more detailed documentation and examples. For example:

- how exactly does _Apply to options -> Value (based field name)_ work?
- how can I use _Use as -> Threshold1_? How can I assign _Threshold1_ a specific color?
- What should you enter in time series at _Field_ -> _Use as_ -> _Time_? Should you just leave it empty? Why is the value at _Select_ set anyway?

**Comment by agross** [comment]
I think I could use this feature to share the max values for two graphs that are located next to each other, but show different data. But I have no idea how I can extract a single "max" value to be used by both graphs from my InfluxDB measurements.

**Comment by savchenko** [comment]
Read the manual, spent 10 minutes trying to configure an elaborate gauge with two integer fields. Clearly I am doing something wrong.

Two fields:
![image](https://user-images.githubusercontent.com/300146/137938368-d0ef3560-7e54-48c8-bc02-a5076697354f.png)

One gauge:
![image](https://user-images.githubusercontent.com/300146/137938411-2c5993f8-fea3-45bb-82aa-32f75dd3220c.png)

"Config query" drop-down returns NIL no matter how many queries are present:
![image](https://user-images.githubusercontent.com/300146/137938674-5c8e15d9-ee11-4535-910b-7b8fcd5adad1.png)

**Comment by ghost** [comment]
As I understand, currently it's possible to use only one "Config from query results" transformation?

When I add a second one the "Config query" dropdown is empty.

I was trying to make a gauge that shows the current cluster CPU usage, has max set to the total CPU count of the cluster and a threshold of 90% of CPU count so:
Query 1 = total_cpu_count = Max
Query 2 = total_cpu_count * 0.9 = Threshold1
Query 3 = current_cpu_usage = Field value

Is this possible or am I doing it wrong?

**Reply by Roemer** [reply]
Did you find out if this is possible now?

**Reply by notemac** [reply]
Did you find out if this is possible now?

**Comment by Cossey** [comment]
In my case, using node exporter I have a metric which has the total memory (`node_memory_MemTotal_bytes{}`) and another separate metric showing the amount active (`node_memory_Active_bytes{}`). I want to use a Gauge, with the Max set to `node_memory_MemTotal_bytes{}` with the value the other metric.

I can't figure out how to do this with *Config from query results*. Am I missing something?

**Comment by offermannu** [comment]
node_memory_MemTotal_bytes{} is your config query which you put into „Config query“ of the „Config from query results“ transformation.
It applies to the „avail“ query. This can be done by choosing „Fields with name“ for the „Apply to“ input field, then you put the field name of the „avail“ query for the „Apply to options“ input field. 
BTW if you put a label in the „legend“ Grafana let you use this label for the „Apply to options“ input field.

In the lower section labeled „Field“ you chose again the „avail“ query and select „max“ from the dropdown select box in the the column labeled „Use as“.

This way the total memory will be used as max value for the gauge.

**Reply by Cossey** [reply]
My man! Thank you. 

Some further notes for others reading:
* Make sure your *Config query* is `Format: Table` on the *Query* tab otherwise it won't show in the *Config from query results* dropdown.
* Make sure you set the *Min* in the Standard Options, if you don't have it in the query, when using a Gauge (I didn't do this and it took me a while to figure out why it was not working).
* If using Prometheus, make sure `Instant` is enabled for the option query.

**Comment by mariotacke** [comment]
@offermannu , does this work with the "Repeat options" functionality? (It doesn't appear to) In my "Config from query results" transform, I can only select the first element of my variables array. The resulting repeat panel shows "no data" for the remaining disks because they are configured to use disk 1's configuration... 🤔 Any ideas?

![image](https://user-images.githubusercontent.com/4942019/143989603-3d1cc839-3d7f-4119-83e6-a4346b2105d8.png)
![image](https://user-images.githubusercontent.com/4942019/143989651-944f6506-1a57-4b61-a279-002a1ac46ffd.png)

**Comment by offermannu** [comment]
Grafana works for me mainly by trial and error, I'm afraid this is more of "error" for me right now. ;-) 
Did you try the repeat option of the dashboard btw?

**Comment by mariotacke** [comment]
I did as well, though it appears to mismatch the template variables and selects the first one by default. My work-around for now is to simply duplicate the configured gauge. Not ideal, but it does the job. Would love for this to work with the repeat option.

**Comment by CatalinMustata** [comment]
This seems very interesting, but I could only have it work with a single config value from a query. 

In my case (running microservices in a Kubernetes cluster), it would be ideal if I could configure multiple standard options from multiple queries - I can't get both CPU requests and limits from a single query, as far as I can see.

**Comment by JDA88** [comment]
I have posted a few scenarios in the issue https://github.com/grafana/grafana/issues/922#issuecomment-1060792920 about using dashboard dynamic variables as threshold in Time series panel with Prometheus queries.

Currently it's not 100% functional, but it's promising, is this the best thread to continue my experiments?

**Comment by justinsmalley** [comment]
This feature has been extremely useful. I would like to see it expanded to support:

- Treat thresholds in a similar manner to value mappings. Have options for "Threshold mapping / Value" and "Threshold mappings / colour", and then allow "All Values" to be selected. This will give a dynamic way to set thresholds. 
- "Soft Min" and "Soft Max" in time series panel.
 
The use case for me is to set the soft min/max a certain amount above or below the threshold lines so that we can always see the lines, but still allow the Y axis to still grow when the values are well above or below the thresholds.

**Reply by aditya-manit** [reply]
Why can't we use it to set dynamic thresholds ?

**Comment by patman-cp** [comment]
FYI, the current documentation for this is here:

https://grafana.com/docs/grafana/next/panels-visualizations/query-transform-data/transform-data/#config-from-query-results

**Comment by unknownowen** [comment]
I am having difficulties using the "Config from query results" option with an InfluxDB query that returns multiple measurements of time series data. I am trying to use tag values metadata stored within the measurement (ex. {unit:kW} to dynamically assign units to each measurement in that database. The database has 15-20 measurements which I currently query all at once using Query A:
"SELECT last(“value”) FROM /./ WHERE time >= now() - 1h and time <= now() GROUP BY time(1s)”

I am trying to config this data by assigning units from a second Query B:
"SHOW tag values FROM /./ WITH key = “unit”"

Doing rows to fields gives me a nice table of the unit string paired with the measurement name. But when I try to do Config from Query it seems to break down. Each measurement instance of Query A is returned simply as "A" and each instance of Query B returned simply as "B". No matter what I pick, I seem to only get the first result displayed as options. It seems like the panel is not updating the field correctly because if I change the query to query a specific measurement, I have to manually disable and refresh the transformation for the field to update with the correct value. Then I can successfully apply units to the one specific measurement. 

what multiple /./ queries look like

![image](https://user-images.githubusercontent.com/118006913/201385666-775cf945-b51b-4f8b-b33a-1da8541f1ba9.png)

![image](https://user-images.githubusercontent.com/118006913/201385797-6c45c385-add3-46d1-bddd-f8a947da32bd.png)

what a single query looks like selecting only /STB1_Power_RL/ (works ok but labor intensive - note that the previous field still shows up as an option, despite the query no longer actually returning - bug?

![image](https://user-images.githubusercontent.com/118006913/201386139-ae9e717f-4fbc-4c7f-b853-9999823ffda0.png)

**Comment by kdorsel** [comment]
Currently when setting up thresholds there is only an option for 1 threshold.
Is it possible to add a threshold2 support? This would allow having a ok zone in the middle and red zone higher than threshold1 and red zone lower than threshold2 for example.
![image](https://user-images.githubusercontent.com/2569415/204053704-6ca6a60d-2d3a-4155-b0f2-e2a394d63d13.png)

The goal is to dynamically do something like this with the two red zones.
![image](https://user-images.githubusercontent.com/2569415/204054233-4bd8678f-ba57-45f0-869a-66e6782941b6.png)

**Comment by notemac** [comment]
Can't use "Config from query results" two or more times in the same panel. There "No options found" when choosing config query.
Please fix it.
![image](https://user-images.githubusercontent.com/43244677/206696012-9b32d944-b04e-40e4-8d07-7c608203cabf.png)

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

We would like all further feedback, whether its issues or feature requests, to go in github issues. As such, I will be closing this discussion.
